File: bindings/python/samples/stream_callback_sample.py
import socket
import socket as _socket
import threading
import time

import zlink


def _tcp_endpoint():
    sock = _socket.socket(_socket.AF_INET, _socket.SOCK_STREAM)
    sock.bind(("127.0.0.1", 0))
    port = sock.getsockname()[1]
    sock.close()
    return port, f"tcp://127.0.0.1:{port}"


def _poll_monitor_event(monitor, timeout_ms):
    with zlink.Poller() as poller:
        poller.add_socket(monitor, zlink.PollEvent.POLLIN)
        ready = poller.poll(timeout_ms)
    if not ready:
        return None
    return monitor.recv()


def _wait_socket_monitor_event(monitor, expected_event, timeout_ms=5000):
    deadline = time.monotonic() + (timeout_ms / 1000.0)
    expected_value = int(expected_event)
    while True:
        remaining_ms = int((deadline - time.monotonic()) * 1000)
        if remaining_ms <= 0:
            raise TimeoutError("socket monitor did not produce the expected event")
        event = _poll_monitor_event(monitor, remaining_ms)
        if event is None:
            continue
        if int(event.event) & expected_value:
            return event


def main():
    port, endpoint = _tcp_endpoint()

    with zlink.Context() as ctx:
        with zlink.StreamSocket(ctx) as server:
            with server.monitor_open(zlink.MonitorEventMask.ACCEPTED) as server_monitor:
                done = threading.Event()
                observed = {}

                def on_message(received):
                    observed["payload"] = received.to_bytes_list()
                    observed["routing_id"] = received.routing_id
                    done.set()

                server.bind(endpoint)
                server.on_receive(on_message)
                with socket.create_connection(("127.0.0.1", port), timeout=3.0) as client:
                    _wait_socket_monitor_event(server_monitor, zlink.MonitorEventMask.ACCEPTED)
                    client.sendall(b"hello-stream")
                    if not done.wait(3.0):
                        raise TimeoutError("stream callback did not receive a message")
                    if observed["payload"] != [b"hello-stream"]:
                        raise AssertionError("unexpected stream callback payload")
                    if not observed["routing_id"]:
                        raise AssertionError("stream callback expected a routing id")
                    server.send(b"hello-stream", routing_id=observed["routing_id"])
                    reply = client.recv(64)
                    if reply != b"hello-stream":
                        raise AssertionError(f"unexpected stream callback reply: {reply!r}")
            print('[stream/callback] send: "hello-stream" → recv: "hello-stream"')


if __name__ == "__main__":
    main()

File: bindings/rust/samples/stream_callback_sample.rs
//! STREAM callback sample – demonstrates STREAM socket with packet handler.
//! The STREAM socket binds as a server; a raw TCP client connects inward.

use std::io::Write;
use std::net::TcpStream;
use std::sync::mpsc;
use std::time::Duration;

use zlink::{Context, SocketMonitor};

pub fn reserve_tcp_port() -> u16 {
    let listener = std::net::TcpListener::bind("127.0.0.1:0").unwrap();
    let port = listener.local_addr().unwrap().port();
    drop(listener);
    port
}

pub fn wait_connected(monitors: &[&zlink::SocketMonitor]) {
    for monitor in monitors {
        loop {
            let event = monitor.recv().expect("monitor recv failed");
            if event.is_connection_ready()
                || monitor
                    .snapshot()
                    .expect("monitor snapshot failed")
                    .is_ready()
            {
                break;
            }
        }
    }
}

pub fn wait_stream_connected(monitor: &zlink::SocketMonitor) {
    loop {
        let event = monitor.recv().expect("monitor recv failed");
        if event.is_accepted()
            || event.is_connection_ready()
            || monitor
                .snapshot()
                .expect("monitor snapshot failed")
                .is_ready()
        {
            break;
        }
    }
}

fn main() {
    let ctx = Context::new().expect("context creation failed");

    let mut stream = ctx.stream_socket().expect("stream socket failed");
    stream.bind("tcp://127.0.0.1:0").expect("bind failed");
    let endpoint = stream.last_endpoint().expect("last_endpoint failed");
    let stream_mon = SocketMonitor::open(&stream).expect("stream monitor open failed");

    let (tx, rx) = mpsc::channel();

    stream
        .on_packet(move |routing_id, header, body| {
            assert!(!routing_id.as_bytes().is_empty());
            assert!(header.as_bytes().is_empty());
            let _ = tx.send(body.as_bytes().to_vec());
        })
        .expect("on_packet failed");

    let tcp_addr = endpoint.strip_prefix("tcp://").unwrap();
    let mut tcp_client = TcpStream::connect(tcp_addr).expect("tcp connect failed");
    tcp_client.set_nodelay(true).expect("set_nodelay failed");
    wait_stream_connected(&stream_mon);
    drop(stream_mon);

    tcp_client
        .write_all(b"hello-stream")
        .expect("tcp write failed");
    tcp_client.flush().expect("tcp flush failed");
    let payload = rx
        .recv_timeout(Duration::from_secs(5))
        .expect("stream callback did not fire within 5s");
    assert_eq!(payload, b"hello-stream");
    let recv_str = std::str::from_utf8(&payload).unwrap();
    println!(
        "[stream/callback] send: \"hello-stream\" → recv: \"{}\"",
        recv_str
    );
}
